src/main/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/service/UserGroupService.kt

```diff
@@ -96,9 +102,17 @@ class UserGroupService(
         userGroupRepository.insertUserGroup(userId, it)
       }
 
-      roleRepository.saveAll(roleRepository.findRolesByGroupCode(group.groupCode).toList())
+      val groupRoles = roleRepository.findAutomaticGroupRolesByGroupCode(groupCode).toSet()
+
+      val existingUserRoles = userRoleService.getUserRoles(userId)
+
+      // remove existing roles to the user
+      val roleCodesToAdd = existingUserRoles?.let { groupRoles.subtract(it.toSet()) }!!.map { it.roleCode }
+
+      userRoleService.addRolesByUserId(userId, roleCodesToAdd)
```

Comment:
I think we did discuss this but wondering if the "distinct" or "distinctBy{it.roleCode} might be nicer/simpler